With factory method overload

alexashley opened the pull request:
Closes #708 

- added `With` overload to `IPostprocessComposer` and updated implementations
-  added unit tests where I saw existing classes being tested (didn't see a test class for the `DelegatingComposer`)

I tried to maintain consistency and add unit tests for all changes, so please let me know where I've erred in that regard.

After these changes, I was able to write this test:

```cs
    public class SomeClass
    {
        public string Property { get; set; }
    }

    public class WithFactoryTest
    {
        [Fact]
        public void ShouldUseFactoryForDynamicBuilder()
        {
            var fixture = new Fixture();
            var callCount = 0;
            Func<string> factory = () =>
            {
                callCount++;
                return $"called {callCount} time{(callCount == 1 ? "" : "s")}";
            };
            var someClassBuilder = fixture.Build<SomeClass>().With(t => t.Property, factory);
            var thingOne = someClassBuilder.Create();
            var thingTwo = someClassBuilder.Create();

            Assert.Equal("called 1 time", thingOne.Property);
            Assert.Equal("called 2 times", thingTwo.Property);
        }
    }

zvirja commented:
Thank you for the contribution.

Before, we proceed with review there are 2 points on which we should agree:
- there is another PR #781 which applies massive changes to the internal implementation of the `NodeComposer.cs`. Likely, we should wait till that PR is merged, so you could rebase your branch and fix all the conflicts.
- You added a [new method](794/files#diff-b96c2ce5ba9e6ba0dca1159d761b2113) to the `IPostProcessComposer` interface, which is a breaking change from the .NET perspective. If anybody has its own interface implementation (which is unlikely, though), the already compiled assemblies will fail with `TypeLoadException` when they update AutoFixture. That is not acceptable according to the [SemVer](http://semver.org/) this project follows. Therefore, likely, we should merge this branch to our next major version - `v4`.

@AutoFixture/core What do you think about the second point from the list above? Do you agree that we cannot merge that to v3 as this is a breaking change?

zvirja commented on Src/AutoFixture/Dsl/CompositeNodeComposer.cs:
```diff
@@ -325,7 +325,35 @@ public IPostprocessComposer<T> OmitAutoProperties()
+        /// further customize the post-processing of created specimens.
+        /// </returns>
+        public IPostprocessComposer<T> With<TProperty>(
+            Expression<Func<T, TProperty>> propertyPicker, Func<TProperty> factory)
```
Please add guard condition to verify that null is not passed (and appropriate test).

zvirja commented on Src/AutoFixture/Kernel/BindingCommand.cs:
```diff
@@ -86,6 +86,22 @@ public BindingCommand(Expression<Func<T, TProperty>> propertyPicker, Func<ISpeci
+        /// A function that creates a value that will be assigned to the property or field
+        /// identified by <paramref name="propertyPicker"/>.
+        /// </param>
+        public BindingCommand(Expression<Func<T, TProperty>> propertyPicker, Func<TProperty> valueCreator)
```
This constructor overload adds no value, as we could use the following one instead:
```csharp
public BindingCommand(Expression<Func<T, TProperty>> propertyPicker, Func<ISpecimenContext, TProperty> valueCreator)
```

You can create a closure and wrap the passed factory here (like `_ => factory()`).
Likely, you created it as we have constructor that receives the value, however I'd vote to not introduce this one.

zvirja commented:
One more concern - could it happen that we'll end up into the situation described [here](https://stackoverflow.com/a/18172472/2009373) that we solved in past? The signatures were exactly the same:
```csharp
Register<T>(this IFixture fixture, T item);
Register<T>(this IFixture fixture, Func<T> creator);
```

VS

```csharp
public IPostprocessComposer<T> With<TProperty>(<Func<T, TProperty>> propertyPicker, TProperty value);
public IPostprocessComposer<T> With<TProperty>(<Func<T, TProperty>> propertyPicker, Func<TProperty> factory);
```

Probably not, as `T` is deduced using the `propertyPicker`, however should be checked.

alexashley commented:
@zvirja I agree that it would be a breaking change, given that the interface is public. 

I'm going to hold off on addressing your comments until #781 merged

moodmosaic commented:
>@AutoFixture/core What do you think about the second point from the list above? Do you agree that we cannot merge that to v3 as this is a breaking change?

Yes, adding new members to an existing, public, interface will break all existing clients, so this should be considered only for doing in v4.

@zvirja, if you ask me, I'd **strongly** suggest to be *extremely* conservative when it comes to modifying types in the *Ploeh.AutoFixture.Dsl* namespace. The types in this namespace has served well for many years, in thousands of unit tests.

If some functionality seems to be missing, I wouldn't rush taking it in as a new feature, at least unless the new members (see #703) know enough about the inner workings of AutoFixture, so they can support it long term, and―at the same time―prevent the codebase from being bloated by each and every hypothetical feature that might come in as a pull request.

It's great that @alexashley sent this pull request, but is it addressing a *common* scenario?

zvirja commented:
> @zvirja, if you ask me, I'd strongly suggest to be extremely conservative when it comes to modifying types in the Ploeh.AutoFixture.Dsl namespace. The types in this namespace have served well for many years, in thousands of unit tests.

Thanks for that valuable input! Probably, I'm too young in this project to clearly feel when it's fine to add feature and when it's better to not extend the codebase. It will take some time till we synchronize our vision 😉

In this particular case I was open to the feature because I discovered that internally we use a factory method, so it will cost very little to add the feature and maintain it later. Also, it seemed to me organic that we have two overloads for the `With()` method - with a static value and a delegate.

> It's great that @alexashley sent this pull request, but is it addressing a common scenario?

To be honest, I stuck with real case scenario here as I never needed such a functionality. But I rarely use customization though. Also I've that the particular case mentioned in #708 might be achieved via extension method.

@alexashley Could you please clarify your real example when you need this feature? Could the extension method suggested [here](https://github.com/AutoFixture/AutoFixture/issues/708#issuecomment-321508595) cover your needs?

zvirja commented:
@moodmosaic Given that you closed #708, we should likely close this one until we again agree to have the feature..

zvirja commented:
@alexashley Thank you for the time you spent on this and sorry for such a decision. Hope you are not frustrated and will continue to contribute to our project. :wink:

alexashley commented:
@zvirja No worries 😄. It was fun to tinker with one of my favorite .NET libraries and I'll keep an eye out for an issue I can pick up.
